sql: store information about keys in orderingInfo, elide distinctNode

RaduBerinde opened the pull request:
#### sql: store information about keys in orderingInfo

We are now able to store information about column sets that form "keys", i.e.
all rows are distinct when projected on that set.

The existing structure had a very limited form of this, only keeping track if
the columns in the ordering form a key. Allowing arbitrary sets will enable more
optimizations.

#### sql: elide distinctNode if we have a key

An immediate application of the new key information is to elide distinct nodes
when possible.

RaduBerinde committed:
sql: store information about keys in orderingInfo

We are now able to store information about column sets that form "keys", i.e.
all rows are distinct when projected on that set.

The existing structure had a very limited form of this, only keeping track if
the columns in the ordering form a key. Allowing arbitrary sets will enable more
optimizations.

Planning benchmark before/after:
```
name                                                             old time/op    new time/op    delta
Planning/SELECT_*_FROM_abc-4                                        214µs ± 5%     212µs ± 3%    ~     (p=0.280 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-4                 193µs ± 8%     186µs ± 5%    ~     (p=0.218 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-4                            186µs ± 6%     185µs ± 3%    ~     (p=0.912 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-4                 186µs ± 5%     192µs ± 9%    ~     (p=0.063 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-4                            182µs ± 3%     196µs ± 6%  +8.12%   (p=0.000 n=7+10)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-4     292µs ± 5%     300µs ± 3%  +2.83%  (p=0.009 n=10+10)

name                                                             old alloc/op   new alloc/op   delta
Planning/SELECT_*_FROM_abc-4                                       31.7kB ± 0%    31.7kB ± 0%  +0.06%   (p=0.026 n=10+9)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-4                22.5kB ± 0%    22.6kB ± 0%  +0.50%    (p=0.000 n=9+9)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-4                           23.4kB ± 0%    23.5kB ± 0%  +0.14%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-4                24.3kB ± 0%    24.4kB ± 0%  +0.54%    (p=0.000 n=9+8)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-4                           23.3kB ± 0%    23.4kB ± 0%  +0.16%  (p=0.022 n=10+10)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-4    71.5kB ± 0%    71.6kB ± 0%  +0.05%  (p=0.000 n=10+10)

name                                                             old allocs/op  new allocs/op  delta
Planning/SELECT_*_FROM_abc-4                                          353 ± 0%       354 ± 0%  +0.34%  (p=0.001 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-4                   363 ± 0%       370 ± 0%  +1.93%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-4                              345 ± 0%       347 ± 0%  +0.58%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-4                   389 ± 0%       397 ± 0%  +2.06%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-4                              341 ± 0%       343 ± 0%  +0.59%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-4       714 ± 0%       716 ± 0%  +0.28%  (p=0.000 n=10+10)
```

RaduBerinde committed:
sql: elide distinctNode if we have a key

An immediate application of the new key information is to elide distinct nodes
when possible.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18508)
<!-- Reviewable:end -->

knz commented:
I'll review this once the first two PRs are merged, so that the diff is more manageable.

RaduBerinde commented:
Rebased, ready to review.

knz commented:
will do

knz commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

As a matter of principle, can you give an estimate in the commit message for the 1st commit of the added allocation overhead for typical (simple) queries? 

---

Reviewed 26 of 26 files at r1, 3 of 4 files at r2.
Review status: 27 of 28 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18508#-:-KvgppfsvrJCM18zQYsr:b-i0141x)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: add some benchmarks targeting the planning process

Add a few simple benchmarks for queries on an empty table, to get running time
and memory allocation information for changes to the planning process.

RaduBerinde commented:
Added some benchmarks for some simple queries on an empty table. We will add more in time.

```
name                                                             old time/op    new time/op    delta
Planning/SELECT_*_FROM_abc-4                                        214µs ± 5%     212µs ± 3%    ~     (p=0.280 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-4                 193µs ± 8%     186µs ± 5%    ~     (p=0.218 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-4                            186µs ± 6%     185µs ± 3%    ~     (p=0.912 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-4                 186µs ± 5%     192µs ± 9%    ~     (p=0.063 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-4                            182µs ± 3%     196µs ± 6%  +8.12%   (p=0.000 n=7+10)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-4     292µs ± 5%     300µs ± 3%  +2.83%  (p=0.009 n=10+10)

name                                                             old alloc/op   new alloc/op   delta
Planning/SELECT_*_FROM_abc-4                                       31.7kB ± 0%    31.7kB ± 0%  +0.06%   (p=0.026 n=10+9)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-4                22.5kB ± 0%    22.6kB ± 0%  +0.50%    (p=0.000 n=9+9)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-4                           23.4kB ± 0%    23.5kB ± 0%  +0.14%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-4                24.3kB ± 0%    24.4kB ± 0%  +0.54%    (p=0.000 n=9+8)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-4                           23.3kB ± 0%    23.4kB ± 0%  +0.16%  (p=0.022 n=10+10)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-4    71.5kB ± 0%    71.6kB ± 0%  +0.05%  (p=0.000 n=10+10)

name                                                             old allocs/op  new allocs/op  delta
Planning/SELECT_*_FROM_abc-4                                          353 ± 0%       354 ± 0%  +0.34%  (p=0.001 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-4                   363 ± 0%       370 ± 0%  +1.93%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-4                              345 ± 0%       347 ± 0%  +0.58%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-4                   389 ± 0%       397 ± 0%  +2.06%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-4                              341 ± 0%       343 ± 0%  +0.59%  (p=0.000 n=10+10)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-4       714 ± 0%       716 ± 0%  +0.28%  (p=0.000 n=10+10)
```

knz commented:
Thanks! gtg :100:
